common.rkt

```diff
@@ -140,23 +140,35 @@
 (define (reify tm st)
   (define index -1)
   (let ((results (let loop ((tm tm) (st st))
-              (define t (walk tm (state-sub st)))
-              (cond ((pair? t) (loop (cdr t) (loop (car t) st)))
-                    ((var? t)  (set! index (+ 1 index))
-                               (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
-                    (else      st)))))
+                   (define t (walk tm (state-sub st)))
+                   (cond ((pair? t) (loop (cdr t) (loop (car t) st)))
+                         ((var? t)  (set! index (+ 1 index))
+                                    (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
+                         (else      st)))))
     (let* ((walked-sub (walk* tm results))
+           (diseq (sort (state-diseq st) (lambda (x y) (< (length x) (length y)))))
```

Comment:
This will compute the correct result.  However, note that computing the `length` of a list is `O(m)` in the length of that list.  With `n` lists of average length `m`, and comparison sorting, we end up with `O(m * n * log n)`.

It's possible to precompute length annotations, sort, then remove the annotations, giving us a smaller `O(m * n + n * log n)` instead.